compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs

```diff
@@ -226,8 +262,11 @@ where
         self.delegate.typing_mode()
     }
 
-    pub(super) fn set_is_normalizes_to_goal(&mut self) {
-        self.is_normalizes_to_goal = true;
+    pub(super) fn step_kind_for_source(&self, source: GoalSource) -> PathKind {
+        match (self.current_goal_kind, source) {
```

Comment:
I wonder if this is a good place to put a comment. Or at least link to one that explains that an impl WC in a coinductive trait counts as a "productive step".